skimage/future/detect_obj/cascade.pyx

```diff
@@ -0,0 +1,832 @@
+        """Search for the object on multiple scales of input image.
+
+        The function takes the input image, the scale factor by which the
+        searching window is multiplied on each step, minimum window size
+        and maximum window size that specify the interval for the search
+        windows that are applied to the input image to detect objects.
+
+        Parameters
+        ----------
+        img : 2-D or 3-D ndarray
+            Ndarray that represents the input image.
+        scale_factor : float
+            The scale by which searching window is multiplied on each step.
+        step_ratio : float
+            The ratio by which the search step in multiplied on each scale
+            of the image. 1 represents the exaustive search and usually is
+            slow. By setting this parameter to higher values the results will
+            be worse but the computation will be much faster. Usually, values
+            in the interval [1, 1.5] give good results.
+        min_size : typle (int, int)
+            Minimum size of the search window.
+        max_size : typle (int, int)
+            Maximum size of the search window.
+        min_neighbour_amount : int
+            Minimum amount of intersecting detections in order for detection
+            to be approved by the function.
+        intersection_score_threshold : float
+            The minimum value of value of ratio
+            (intersection area) / (small rectangle ratio) in order to merge
+            two detections into one.
+
+        Returns
+        -------
+        output : list of dicts
+            Dict have form {'r': int, 'c': int, 'width': int, 'height': int},
+            where 'r' represents row position of top left corner of detected
+            window, 'c' - col position, 'width' - width of detected window,
+            'height' - height of detected window.
+        """
+
+        cdef:
+            Py_ssize_t max_row
+            Py_ssize_t max_col
+            Py_ssize_t current_height
+            Py_ssize_t current_width
+            Py_ssize_t current_row
+            Py_ssize_t current_col
+            Py_ssize_t current_step
+            Py_ssize_t amount_of_scales
+            Py_ssize_t img_height
+            Py_ssize_t img_width
+            Py_ssize_t scale_number
+            Py_ssize_t window_height = self.window_height
+            Py_ssize_t window_width = self.window_width
+            int result
+            float[::1] scale_factors
+            float[:, ::1] int_img
+            float current_scale_factor
+            vector[Detection] output
+            Detection new_detection
+
+        int_img = self._get_contiguous_integral_image(img)
+        img_height = int_img.shape[0]
+        img_width = int_img.shape[1]
+
+        scale_factors = self._get_valid_scale_factors(min_size, max_size, scale_factor)
+        amount_of_scales = scale_factors.shape[0]
+
+        # Initialize lock to enable thread-safe writes to the array
```

Comment:
@stefanv Yes it is, `with` won't make a single `vector` thread-safe. The alternative is to use an array of vectors and no locks, but this solution is much cleaner.

dipy uses `prange` with some safety checks
https://github.com/nipy/dipy/blob/master/dipy/align/bundlemin.pyx

scikit-learn currently chooses not to use it
https://github.com/scikit-learn/scikit-learn/pull/4873

But I think we can explicitly document the limitations, and keep the implementation in `future`. I am specially tempted because of how exclusive detection at each scale is.